vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+            no_duplication = false;
+        }
+    }
+
+    return no_duplication;
+}
+
+/* Check the whole Routing Resource Graph  
+ * identify and report any duplicated edges between two nodes 
+ */
+bool RRGraph::check_duplicated_edges() const {
+    bool no_duplication = true;
+    /* For each node:
+     * Search input edges, see there are two edges with same id or address 
+     */
+    for (auto node : nodes()) {
+        if (false == check_node_duplicated_edges(node)) {
+            no_duplication = false;
+        }
+    }
+
+    return no_duplication;
+}
+
+/* 
+ * identify and report any duplicated edges between two nodes 
+ */
+bool RRGraph::check_dangling_nodes() const {
+    bool no_dangling = true;
+    /* For each node: 
+     * check if the number of input edges and output edges are both 0
+     * If so, this is a dangling nodes and report 
+     */
+    for (auto node : nodes()) {
+        if ((0 == this->node_fan_in(node))
+            && (0 == this->node_fan_out(node))) {
+            /* Print a warning! */
+            VTR_LOG_WARN("Node %s is dangling (zero fan-in and zero fan-out)!\n",
+                         node);
+            VTR_LOG_WARN("Node details for debugging:\n");
+            this->print_node(node);
+            no_dangling = false;
+        }
+    }
+
+    return no_dangling;
+}
+
+/* This function should be used when nodes, edges, switches and segments have been used after 
+ * We will build the fast_lookup, partition edges and check 
+ */
+bool RRGraph::check() const {
+    bool check_flag = true;
+    size_t num_err = 0;
+
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+
+    /* Fundamental check */
+    if (false == check_nodes_edges()) {
+        VTR_LOG_WARN("Fail in checking edges connected to each node!\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    if (false == check_node_segments()) {
+        VTR_LOG_WARN("Fail in checking segment IDs of nodes !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    if (false == check_edge_src_nodes()) {
+        VTR_LOG_WARN("Fail in checking source nodes of edges !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    if (false == check_edge_sink_nodes()) {
+        VTR_LOG_WARN("Fail in checking sink nodes of edges !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    if (false == check_edge_switches()) {
+        VTR_LOG_WARN("Fail in checking switch IDs of edges !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    /* Advanced check */
+    if (false == check_duplicated_edges()) {
+        VTR_LOG_WARN("Fail in checking duplicated edges !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    if (false == check_dangling_nodes()) {
+        VTR_LOG_WARN("Fail in checking dangling nodes !\n");
+        check_flag = false;
+        num_err++;
+    }
+
+    VTR_LOG("Checked Routing Resource graph with %d errors !\n",
```

Comment:
Should think carefully about whether this should be a warning or a hard error (i.e. throw exception with VPR_ERROR).

We likely also want to subdivide these checks into two categories:
 1. Things that indicate the data structure is inconsistent
 2. Other issues which may be odd or indicate potentially incorrect usage of the data structure

I think that anything that falls into category (1) should be a VPR_ERROR, halting execution. If the data structures are inconsistent (e.g. an edge's src/sink node disagrees with the node's edge) then there is nothing we can do and we should die instead of continuing with corrupt/inconsistent data.

Things in categeory (2) we can issue warnings for. Although we should consider whether they are fundamental things which are assumed by the RR Graph object data structure (in which case they should be checked here), or whether they are assumptions which are not fundamental to the data structure itself, and can be checked using the external public interface of the RR graph object (such things would usually include things that VPR assumes about the RR graphs it is working on, and should be in VPR's check_rr_graph()).

Reply:
I have categorized the first type of errors to fatal one, leading VPR to halt.
It means the RRGraph goes wrong and routers will fail. 
The second type of error are optional. The graph is still routable but these errors could be potential sources of bugs.